File: src/discord_qc/discord/interactions/command.clj
(ns discord-qc.discord.interactions.command
  (:require [clojure.string :as string :refer [lower-case]]
            [clojure.set :as set]

            [discljord.messaging :as discord-rest]

            [slash.response :as srsp]
            [slash.component.structure :as scomp]

            [com.rpl.specter :as s]

            [taoensso.timbre :as timbre :refer [log]]

            [discord-qc.state :refer [state* discord-state*]]
            [discord-qc.elo :as elo]
            [discord-qc.handle-db :as db]
            [discord-qc.handle-gh :as gh]
            [discord-qc.utils :refer [get-keys-starting-with]]
            [discord-qc.discord.utils :refer [get-voice-channel-members
                                              user-in-voice-channel?
                                              build-components-action-rows
                                              get-user-display-name]]
            [discord-qc.discord.commands :refer [application-commands admin-commands owner-commands sorting-options]]
            [discord-qc.discord.interactions.utils :refer [map-command-interaction-options
                                                           divide-hub]]))

(defn elo-map->embed [elo-map]
  (let [format-map-entry (fn [[game-mode mode-score]] (hash-map :name (game-mode elo/mode-names)
                                                                :value (str (format "%.3f" (double mode-score)))))]
    [{:type "rich" :title "Quake ELO for:" :description (:quake-name elo-map) :color 9896156
      :fields (concat
               (map format-map-entry (select-keys elo-map (keys elo/mode-names))))}]))

(defn clean-user-id [user-id]
  (->> user-id
       (filter #(not (contains? #{\@ \> \<} %)))
       (apply str)))

;; Command interactions
(defmulti handle-command-interaction
  (fn [interaction] (get-in interaction [:data :name])))

(defmethod handle-command-interaction "query" [interaction]
  (let [discord-id (clean-user-id (get (map-command-interaction-options interaction) "discord-id"))]

    (if-let [elo (elo/discord-id->Elo discord-id)]
      (do
        (srsp/update-message {:content "" :embeds (elo-map->embed elo)}))
      (srsp/update-message {:content (str "couldn't find data for user")}))))

(defmethod handle-command-interaction "rename" [interaction]
  (let [interaction-options (map-command-interaction-options interaction)
        discord-id (s/select-first [:member :user :id] interaction)
        quake-name (get interaction-options "quake-name")]

    (if-let [elo (elo/discord-id->Elo discord-id)]
      (do
        (let [new-elo (assoc elo :quake-name quake-name)]
          (elo/save-discord-id->Elo discord-id new-elo)
          (srsp/update-message {:content "" :embeds (elo-map->embed new-elo)})))
      (srsp/update-message {:content (str "couldn't find user")}))))

(defmethod handle-command-interaction "balance" [interaction]
  (let [interaction-options (map-command-interaction-options interaction)
        game-mode (get interaction-options "game-mode")
        manual-entries (->> (get-keys-starting-with interaction-options "player-tag")
                            (vals)
                            (map clean-user-id))

        guild-id (:guild-id interaction)
        interactor-id (s/select-first [:member :user :id] interaction)

        voice-channel-id (user-in-voice-channel? interactor-id)
        voice-channel-members (get-voice-channel-members voice-channel-id)

        found-players (into #{} (concat manual-entries voice-channel-members))

        elos (map #(assoc (elo/discord-id->Elo %) :discord-id %) found-players)

        unregistered-users (s/select [s/ALL #(not (:quake-name %)) :discord-id] elos)
        unregistered-users-names (into {} (map #(hash-map % (get-user-display-name guild-id %)) unregistered-users))

        quake-name-button (fn [elo] (let [discord-id (:discord-id elo)]
                                      (scomp/button :secondary
                                                    (str "toggle-primary-secondary/" discord-id)
                                                    :label (get elo :quake-name (get unregistered-users-names discord-id)))))
        components (build-components-action-rows
                    (concat
                     (map quake-name-button elos)
                     [(scomp/button :danger  "select-all-primary-secondary" :label "Select All")
                      (scomp/button :success  (str "balance!/" game-mode) :label "Balance!")]))
        content (string/join "\n"
                             [(when (not-empty unregistered-users-names)
                                (str "Unregistered Users: " (string/join ", " (vals unregistered-users-names))))
                              (str "Balancing for " game-mode)])]
    (srsp/channel-message {:content content :components components})))

(defmethod handle-command-interaction "divide" [interaction]
  (let [interaction-options (map-command-interaction-options interaction)
        guild-id (:guild-id interaction)
        user-id (s/select-first [:member :user :id] interaction)
        game-mode (get (set/map-invert elo/mode-names) (get interaction-options "game-mode"))
        sorting-method (get interaction-options "sort-by" (-> sorting-options (first) (:value)))

        ignored-players (->> (get-keys-starting-with interaction-options "spectator-tag")
                             (vals)
                             (map clean-user-id)
                             (set))

        manual-entries (->> (get-keys-starting-with interaction-options "player-tag")
                            (vals)
                            (map clean-user-id)
                            (set))]

    (srsp/channel-message (divide-hub (str guild-id)
                                      (str user-id)
                                      game-mode
                                      sorting-method
                                      manual-entries
                                      ignored-players))))

;; Admin commands

(defmethod handle-command-interaction "register" [interaction]
  (let [interaction-options (map-command-interaction-options interaction)
        discord-id (clean-user-id (get interaction-options "discord-id"))
        quake-name (get interaction-options "quake-name")
        score (get interaction-options "score")]

    (let [elo-map (elo/create-elo-map quake-name score)]
      (elo/save-discord-id->Elo discord-id elo-map)
      (srsp/update-message {:content "" :embeds (elo-map->embed elo-map)}))))

(defmethod handle-command-interaction "rename-other" [interaction]
  (let [interaction-options (map-command-interaction-options interaction)
        discord-id (clean-user-id (get interaction-options "discord-id"))
        quake-name (get interaction-options "quake-name")]

    (if-let [elo (elo/discord-id->Elo discord-id)]
      (do
        (let [new-elo (assoc elo :quake-name quake-name)]
          (elo/save-discord-id->Elo discord-id new-elo)
          (srsp/update-message {:content "" :embeds (elo-map->embed new-elo)})))
      (srsp/update-message {:content (str "couldn't find user")}))))

(defmethod handle-command-interaction "adjust" [interaction]
  (let [interaction-options (map-command-interaction-options interaction)
        discord-id (clean-user-id (get interaction-options "discord-id"))
        mode (get (set/map-invert elo/mode-names) (get interaction-options "game-mode"))
        score (get interaction-options "score")]

    (if-let [elo (elo/discord-id->Elo discord-id)]
      (do
        (let [new-elo (assoc elo mode score)]
          (elo/save-discord-id->Elo discord-id new-elo)
          (srsp/update-message {:content "" :embeds (elo-map->embed new-elo)})))
      (srsp/update-message {:content (str "couldn't find user")}))))

(defmethod handle-command-interaction "db-stats" [interaction]
  (let [players-registered @db/all-discord-ids-in-db*
        db-stats-message (string/join "\n"
                                      [(str "\\# Of players registered in db: " (count players-registered))])]

    (srsp/channel-message {:content db-stats-message})))

(defmethod handle-command-interaction "make-admin" [interaction]
  (let [interaction-options (map-command-interaction-options interaction)
        discord-id (clean-user-id (get interaction-options "discord-id"))]

    (db/save-admin-id discord-id)
    (srsp/channel-message {:content (str "User " discord-id " is now an admin")})))

(defmethod handle-command-interaction "list-admins" [interaction]
  (let [interaction-options (map-command-interaction-options interaction)
        admin-ids @db/admin-ids*
        admin-names (->> admin-ids
                         (map #(deref (discord-rest/get-user! (:rest @state*) %)))
                         (map #(or (:global-name %) (:username %))))]

    (srsp/channel-message {:content (string/join ", " admin-names)})))

(defmethod handle-command-interaction "backup-db" [interaction]
  (let [interaction-options (map-command-interaction-options interaction)
        db-edn (db/db->edn)]
    (gh/update-db-file db-edn)
    (srsp/channel-message {:content "https://raw.githubusercontent.com/barakor/discord_qc/db-data/db-data.edn"})))

(defmethod handle-command-interaction "restore-db-backup" [interaction]
  (let [interaction-options (map-command-interaction-options interaction)]
    (db/refresh-db-from-github)
    (srsp/channel-message {:content "DB restored from backup"})))

(defonce ^:private admin-only-commands (set (map :name admin-commands)))

(defonce ^:private owner-only-commands (set (map :name owner-commands)))

(defonce ^:private application-commands-names (set (map :name application-commands)))

(defn execute-interaction [interaction]
  (let [{:keys [type data]} (handle-command-interaction interaction)
        edit-response @(discord-rest/edit-original-interaction-response! (:rest @state*) (:application-id interaction) (:token interaction) data)]
    (log :debug edit-response)))

(defn command-interaction [interaction]
  @(discord-rest/create-interaction-response! (:rest @state*) (:id interaction) (:token interaction) (:type srsp/deferred-channel-message))
  (try
    (let [command-interaction-name (get-in interaction [:data :name])
          interactor-id (s/select-first [:member :user :id] interaction)]
      (cond
        ;; owner commands
        (and (contains? owner-only-commands command-interaction-name)
             (= "88533822521507840" interactor-id)) (execute-interaction interaction)

        ;; admin commands
        (and (contains? admin-only-commands command-interaction-name)
             (contains? @db/admin-ids* interactor-id)) (execute-interaction interaction)

        ;; other commands
        (contains? application-commands-names command-interaction-name) (execute-interaction interaction)

        ;; else fail
        :else @(discord-rest/edit-original-interaction-response! (:rest @state*)
                                                                 (:application-id interaction)
                                                                 (:token interaction)
                                                                 {:content "You are not authorized to use this command"})))
    (catch Exception e (do (log :error e)
                           (@(discord-rest/edit-original-interaction-response! (:rest @state*)
                                                                               (:application-id interaction)
                                                                               (:token interaction)
                                                                               {:content "Failed to process command"}))))))

File: src/interactions/command.rs
use crate::{
    balancing::SortMethod,
    config_handler::GithubConfig,
    db_handler::{Db, GameMode, PlayerElo, player_elo_embed},
    discord_utils::{
        build_components_action_rows, button, str_to_id, text_response, trim_tags,
        user_voice_channel, voice_channel_members,
    },
    event_handler::Bot,
    github_handler::{get_bytes_from_github, upload_bytes_to_github},
    interactions::utils::{divide_hub, named_elos},
};
use anyhow::{Context, Result, anyhow};
use std::{collections::BTreeSet, sync::Arc};
use tokio::sync::RwLock;
use twilight_interactions::command::{CommandModel, CommandOption, CreateCommand, CreateOption};
use twilight_model::{
    application::interaction::{
        Interaction,
        application_command::{CommandData, CommandDataOption, CommandOptionValue},
    },
    channel::message::component::ButtonStyle,
    http::interaction::InteractionResponseData,
    id::{Id, marker::UserMarker},
};
use twilight_util::builder::InteractionResponseDataBuilder;

/// Authorization tier required to invoke a command.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Permission {
    /// Anyone may use it.
    App,
    /// Holders of the admin role in the home server (and the owner).
    Admin,
    /// Bot owner only.
    Owner,
}

/// Compile-time equality of two `&str` (`==` isn't const for strings yet).
const fn str_eq(a: &str, b: &str) -> bool {
    let (a, b) = (a.as_bytes(), b.as_bytes());
    if a.len() != b.len() {
        return false;
    }
    let mut i = 0;
    while i < a.len() {
        if a[i] != b[i] {
            return false;
        }
        i += 1;
    }
    true
}

/// The single source of truth for every slash command. Each row is:
///
/// ```text
/// Variant => StructType, "wire-name", Permission, ephemeral: bool, mutating: bool;
/// ```
///
/// From this one table the macro generates the `Command` enum and all of its
/// metadata accessors (`name`, `from_name`, `permission`, `is_ephemeral`,
/// `is_mutating`), plus a per-row compile-time assertion that the wire name
/// matches the struct's `#[command(name = ...)]` (via `CreateCommand::NAME`).
///
/// Adding a command means adding ONE row here. Its name check is generated
/// automatically, and the exhaustive `match` in `handle_command` then fails to
/// compile until you wire the handler — so a new command can't be half-added.
macro_rules! commands {
    ($(
        $variant:ident => $struct:ty, $name:literal, $perm:ident,
        ephemeral: $eph:literal, mutating: $mutating:literal
    );* $(;)?) => {
        /// Every slash command the bot handles.
        #[derive(Debug, Clone, Copy, PartialEq, Eq)]
        pub enum Command {
            $($variant),*
        }

        impl Command {
            /// Discord wire name, matching each struct's `#[command(name = ...)]`.
            pub const fn name(self) -> &'static str {
                match self {
                    $(Command::$variant => $name),*
                }
            }

            /// Parse an incoming interaction's command name.
            pub fn from_name(name: &str) -> Option<Self> {
                Some(match name {
                    $($name => Command::$variant,)*
                    _ => return None,
                })
            }

            /// Authorization tier required to run this command.
            pub fn permission(self) -> Permission {
                match self {
                    $(Command::$variant => Permission::$perm),*
                }
            }

            /// Response is only shown to the invoking user.
            pub fn is_ephemeral(self) -> bool {
                match self {
                    $(Command::$variant => $eph),*
                }
            }

            /// Mutates the db; success triggers a persist + github backup.
            pub fn is_mutating(self) -> bool {
                match self {
                    $(Command::$variant => $mutating),*
                }
            }
        }

        // Build fails if any row's wire name drifts from the struct's macro name.
        const _: () = {
            $(assert!(str_eq(<$struct>::NAME, Command::$variant.name()));)*
        };
    };
}

commands! {
    Balance         => BalanceCommand,         "balance",           App,   ephemeral: false, mutating: false;
    Divide          => DivideCommand,          "divide",            App,   ephemeral: false, mutating: false;
    Query           => QueryCommand,           "query",             Admin, ephemeral: true,  mutating: false;
    Rename          => RenameCommand,          "rename",            App,   ephemeral: true,  mutating: true;
    RenameOther     => RenameOtherCommand,     "rename-other",      Admin, ephemeral: true,  mutating: true;
    Register        => RegisterCommand,        "register",          Admin, ephemeral: true,  mutating: true;
    Adjust          => AdjustCommand,          "adjust",            Admin, ephemeral: true,  mutating: true;
    DbStats         => DBStatsCommand,         "db-stats",          Admin, ephemeral: true,  mutating: false;
    BackupDb        => BackupDbCommand,        "backup-db",         Owner, ephemeral: true,  mutating: true;
    RestoreDbBackup => RestoreDBBackupCommand, "restore-db-backup", Owner, ephemeral: true,  mutating: true;
}

/// Render a slash command invocation back to its textual form, e.g.
/// `/register discord_id:<@123> quake_name:foo score:5`. Used for the bot-logs
/// audit line so the full command (name + every supplied option) is captured.
pub fn render_invocation(data: &CommandData) -> String {
    let mut out = format!("/{}", data.name);
    render_options(&data.options, &mut out);
    out
}

fn render_options(options: &[CommandDataOption], out: &mut String) {
    for option in options {
        match &option.value {
            CommandOptionValue::SubCommand(sub) | CommandOptionValue::SubCommandGroup(sub) => {
                out.push(' ');
                out.push_str(&option.name);
                render_options(sub, out);
            }
            value => {
                out.push_str(&format!(" {}:{}", option.name, render_value(value)));
            }
        }
    }
}

fn render_value(value: &CommandOptionValue) -> String {
    match value {
        CommandOptionValue::String(s) => s.clone(),
        CommandOptionValue::Integer(i) => i.to_string(),
        CommandOptionValue::Number(n) => n.to_string(),
        CommandOptionValue::Boolean(b) => b.to_string(),
        CommandOptionValue::User(id) => format!("<@{}>", id),
        CommandOptionValue::Mentionable(id) => format!("<@{}>", id),
        CommandOptionValue::Channel(id) => format!("<#{}>", id),
        CommandOptionValue::Role(id) => format!("<@&{}>", id),
        CommandOptionValue::Attachment(id) => id.to_string(),
        CommandOptionValue::Focused(s, _) => s.clone(),
        CommandOptionValue::SubCommand(_) | CommandOptionValue::SubCommandGroup(_) => String::new(),
    }
}

/// Game modes exposed as slash command choices (mirrors the Clojure choice list).
#[derive(CommandOption, CreateOption, Debug, Clone, Copy)]
pub enum GameModeOption {
    #[option(name = "Sacrifice", value = "sacrifice")]
    Sacrifice,
    #[option(name = "Objective", value = "objective")]
    Objective,
    #[option(name = "Killing", value = "killing")]
    Killing,
    #[option(name = "Sacrifice Tournament", value = "sacrifice-tournament")]
    SacrificeTournament,
    #[option(name = "Slipgate", value = "slipgate")]
    Slipgate,
    #[option(name = "CTF", value = "ctf")]
    Ctf,
    #[option(name = "TDM", value = "tdm")]
    Tdm,
    #[option(name = "TDM 2V2", value = "tdm-2v2")]
    Tdm2v2,
}

impl From<GameModeOption> for GameMode {
    fn from(option: GameModeOption) -> Self {
        match option {
            GameModeOption::Sacrifice => GameMode::Sacrifice,
            GameModeOption::Objective => GameMode::Objective,
            GameModeOption::Killing => GameMode::Killing,
            GameModeOption::SacrificeTournament => GameMode::SacrificeTournament,
            GameModeOption::Slipgate => GameMode::Slipgate,
            GameModeOption::Ctf => GameMode::Ctf,
            GameModeOption::Tdm => GameMode::Tdm,
            GameModeOption::Tdm2v2 => GameMode::Tdm2v2,
        }
    }
}

#[derive(CommandModel, CreateCommand, Debug)]
#[command(name = "query", desc = "Query Quake player's stats")]
pub struct QueryCommand {
    #[command(desc = "Discord Tag/ID or Quake Name")]
    pub quaker: String,
}

impl QueryCommand {
    pub async fn handle(
        data: CommandData,
        db: &Arc<RwLock<Db>>,
    ) -> Result<Option<InteractionResponseData>> {
        let command =
            QueryCommand::from_interaction(data.into()).context("failed to parse command data")?;

        let trimmed_quaker = trim_tags(&command.quaker);
        let db = db.read().await;
        let user = match trimmed_quaker.parse::<u64>() {
            Ok(discord_id) => db.elos.get(&discord_id),
            Err(_) => db.by_quake_name(trimmed_quaker),
        };
        match user {
            Some(user) => Ok(Some(player_elo_embed(&user))),
            None => Ok(Some(text_response("couldn't find data for user"))),
        }
    }
}

#[derive(CommandModel, CreateCommand, Debug)]
#[command(name = "rename", desc = "Rename Quaker")]
pub struct RenameCommand {
    #[command(desc = "Quake Name")]
    pub quake_name: String,
}

impl RenameCommand {
    pub async fn handle(
        data: CommandData,
        db: &Arc<RwLock<Db>>,
        user_id: u64,
        log_detail: &mut Option<String>,
    ) -> Result<Option<InteractionResponseData>> {
        let command =
            RenameCommand::from_interaction(data.into()).context("failed to parse command data")?;

        let new_name = command.quake_name;
        let mut db = db.write().await;
        match db.rename(user_id, new_name.clone()) {
            Some(old_name) => {
                *log_detail = Some(format!("{old_name} -> {new_name}"));
                Ok(Some(
                    InteractionResponseDataBuilder::new()
                        .content(format!("Renamed to `{new_name}`"))
                        .build(),
                ))
            }
            None => Ok(Some(text_response("couldn't find user"))),
        }
    }
}

#[derive(CommandModel, CreateCommand, Debug)]
#[command(name = "rename-other", desc = "Rename Quaker")]
pub struct RenameOtherCommand {
    #[command(desc = "Tag a discord user")]
    pub discord_id: Id<UserMarker>,
    #[command(desc = "Quake Name")]
    pub quake_name: String,
}

impl RenameOtherCommand {
    pub async fn handle(
        data: CommandData,
        db: &Arc<RwLock<Db>>,
        log_detail: &mut Option<String>,
    ) -> Result<Option<InteractionResponseData>> {
        let command = RenameOtherCommand::from_interaction(data.into())
            .context("failed to parse command data")?;

        let discord_id = command.discord_id.get();
        let new_name = command.quake_name;
        let mut db = db.write().await;
        match db.rename(discord_id, new_name.clone()) {
            Some(old_name) => {
                *log_detail = Some(format!("{old_name} -> {new_name}"));
                let elo = db.elos.get(&discord_id).expect("just renamed, must exist");
                Ok(Some(player_elo_embed(elo)))
            }
            None => Ok(Some(text_response("couldn't find user"))),
        }
    }
}

#[derive(CommandModel, CreateCommand, Debug)]
#[command(name = "register", desc = "Register Quaker")]
pub struct RegisterCommand {
    #[command(desc = "Tag a discord user")]
    pub discord_id: Id<UserMarker>,
    #[command(desc = "Quake Name")]
    pub quake_name: String,
    #[command(desc = "Player's score for all modes")]
    pub score: f64,
}

impl RegisterCommand {
    pub async fn handle(
        data: CommandData,
        db: &Arc<RwLock<Db>>,
    ) -> Result<Option<InteractionResponseData>> {
        let command = RegisterCommand::from_interaction(data.into())
            .context("failed to parse command data")?;

        let discord_id = command.discord_id.get();
        let elo = PlayerElo::with_score(command.quake_name, command.score);
        db.write().await.register(discord_id, elo.clone());
        Ok(Some(player_elo_embed(&elo)))
    }
}

#[derive(CommandModel, CreateCommand, Debug)]
#[command(name = "adjust", desc = "Adjust Quaker Mode Score")]
pub struct AdjustCommand {
    #[command(desc = "Tag a discord user")]
    pub discord_id: Id<UserMarker>,
    #[command(desc = "Player's score for the mode")]
    pub score: f64,
    #[command(desc = "Game Mode (defaults to Sacrifice Tournament)")]
    pub game_mode: Option<GameModeOption>,
}

impl AdjustCommand {
    pub async fn handle(
        data: CommandData,
        db: &Arc<RwLock<Db>>,
        log_detail: &mut Option<String>,
    ) -> Result<Option<InteractionResponseData>> {
        let command =
            AdjustCommand::from_interaction(data.into()).context("failed to parse command data")?;

        let discord_id = command.discord_id.get();
        let mode = command
            .game_mode
            .unwrap_or(GameModeOption::SacrificeTournament)
            .into();
        let mut db = db.write().await;
        match db.elos.get_mut(&discord_id) {
            Some(elo) => {
                *log_detail = Some(format!("{} -> {}", elo.score(mode), command.score));
                elo.set_score(mode, command.score);
                Ok(Some(player_elo_embed(elo)))
            }
            None => Ok(Some(text_response("couldn't find user"))),
        }
    }
}

#[derive(CommandModel, CreateCommand, Debug)]
#[command(name = "db-stats", desc = "Get stats about the db")]
pub struct DBStatsCommand {}

impl DBStatsCommand {
    pub async fn handle(db: &Arc<RwLock<Db>>) -> Result<Option<InteractionResponseData>> {
        let players_registered = db.read().await.elos.len();
        Ok(Some(text_response(format!(
            "# Of players registered in db: {}",
            players_registered
        ))))
    }
}

/// Sort orders exposed on /divide.
#[derive(CommandOption, CreateOption, Debug, Clone, Copy)]
pub enum SortMethodOption {
    #[option(name = "Random", value = "random")]
    Random,
    #[option(name = "Player's Score", value = "score")]
    Score,
}

impl From<SortMethodOption> for SortMethod {
    fn from(option: SortMethodOption) -> Self {
        match option {
            SortMethodOption::Random => SortMethod::Random,
            SortMethodOption::Score => SortMethod::Score,
        }
    }
}

/// Parse the optional player/spectator tag options into a set of ids,
/// silently skipping anything that isn't a user mention.
fn tag_ids(tags: [&Option<String>; 8]) -> BTreeSet<u64> {
    tags.iter()
        .filter_map(|tag| tag.as_deref())
        .filter_map(|tag| str_to_id(tag).ok())
        .collect()
}

#[derive(CommandModel, CreateCommand, Debug)]
#[command(name = "balance", desc = "Balance a Quake Champions lobby")]
pub struct BalanceCommand {
    #[command(desc = "Game Mode (defaults to Sacrifice Tournament)")]
    pub game_mode: Option<GameModeOption>,
    #[command(desc = "Manually add tagged discord user to lobby")]
    pub player_tag1: Option<String>,
    #[command(desc = "Manually add tagged discord user to lobby")]
    pub player_tag2: Option<String>,
    #[command(desc = "Manually add tagged discord user to lobby")]
    pub player_tag3: Option<String>,
    #[command(desc = "Manually add tagged discord user to lobby")]
    pub player_tag4: Option<String>,
    #[command(desc = "Manually add tagged discord user to lobby")]
    pub player_tag5: Option<String>,
    #[command(desc = "Manually add tagged discord user to lobby")]
    pub player_tag6: Option<String>,
    #[command(desc = "Manually add tagged discord user to lobby")]
    pub player_tag7: Option<String>,
    #[command(desc = "Manually add tagged discord user to lobby")]
    pub player_tag8: Option<String>,
}

impl BalanceCommand {
    /// Post the player-selection message: one toggle button per player found
    /// in the caller's voice channel or tagged manually, plus Select All and
    /// Balance! triggers handled as component interactions.
    pub async fn handle(
        data: CommandData,
        bot: &Bot,
        interaction: &Interaction,
    ) -> Result<Option<InteractionResponseData>> {
        let command = BalanceCommand::from_interaction(data.into())
            .context("failed to parse command data")?;
        let guild_id = interaction
            .guild_id
            .ok_or(anyhow!("balance outside a guild"))?;
        let user_id = interaction
            .author_id()
            .ok_or(anyhow!("interaction without author"))?;
        let game_mode: GameMode = command
            .game_mode
            .unwrap_or(GameModeOption::SacrificeTournament)
            .into();

        let manual_entries = tag_ids([
            &command.player_tag1,
            &command.player_tag2,
            &command.player_tag3,
            &command.player_tag4,
            &command.player_tag5,
            &command.player_tag6,
            &command.player_tag7,
            &command.player_tag8,
        ]);

        let members = user_voice_channel(&bot.cache, guild_id, user_id)
            .map(|channel_id| voice_channel_members(&bot.cache, channel_id))
            .unwrap_or_default();

        let found_players: BTreeSet<u64> = manual_entries
            .iter()
            .chain(members.iter())
            .copied()
            .collect();

        let (players, unregistered_names) =
            named_elos(bot, guild_id, game_mode, &found_players).await;

        let mut buttons: Vec<_> = players
            .iter()
            .map(|player| {
                button(
                    ButtonStyle::Secondary,
                    format!("toggle-primary-secondary/{}", player.id),
                    player.name.clone(),
                )
                .into()
            })
            .collect();
        buttons.push(
            button(
                ButtonStyle::Danger,
                "select-all-primary-secondary",
                "Select All",
            )
            .into(),
        );
        buttons.push(
            button(
                ButtonStyle::Success,
                format!("balance!/{}", game_mode.name()),
                "Balance!",
            )
            .into(),
        );

        let mut content_lines = Vec::new();
        if !unregistered_names.is_empty() {
            content_lines.push(format!(
                "Unregistered Users: {}",
                unregistered_names.join(", ")
            ));
        }
        content_lines.push(format!("Balancing for {}", game_mode.name()));

        Ok(Some(
            InteractionResponseDataBuilder::new()
                .content(content_lines.join("\n"))
                .components(build_components_action_rows(buttons))
                .build(),
        ))
    }
}

#[derive(CommandModel, CreateCommand, Debug)]
#[command(name = "divide", desc = "Divide hub inhabitants to other lobbies")]
pub struct DivideCommand {
    #[command(desc = "Game Mode (defaults to Sacrifice Tournament)")]
    pub game_mode: Option<GameModeOption>,
    #[command(desc = "Sort players before dividing them")]
    pub sort_by: Option<SortMethodOption>,
    #[command(desc = "Manually tag discord user as a spectator")]
    pub spectator_tag1: Option<String>,
    #[command(desc = "Manually tag discord user as a spectator")]
    pub spectator_tag2: Option<String>,
    #[command(desc = "Manually tag discord user as a spectator")]
    pub spectator_tag3: Option<String>,
    #[command(desc = "Manually tag discord user as a spectator")]
    pub spectator_tag4: Option<String>,
    #[command(desc = "Manually add tagged discord user to lobby")]
    pub player_tag1: Option<String>,
    #[command(desc = "Manually add tagged discord user to lobby")]
    pub player_tag2: Option<String>,
    #[command(desc = "Manually add tagged discord user to lobby")]
    pub player_tag3: Option<String>,
    #[command(desc = "Manually add tagged discord user to lobby")]
    pub player_tag4: Option<String>,
}

impl DivideCommand {
    pub async fn handle(
        data: CommandData,
        bot: &Bot,
        interaction: &Interaction,
    ) -> Result<Option<InteractionResponseData>> {
        let command =
            DivideCommand::from_interaction(data.into()).context("failed to parse command data")?;
        let guild_id = interaction
            .guild_id
            .ok_or(anyhow!("divide outside a guild"))?;
        let user_id = interaction
            .author_id()
            .map(|id| id.get())
            .ok_or(anyhow!("interaction without author"))?;

        let sort_method: SortMethod = command.sort_by.unwrap_or(SortMethodOption::Random).into();
        let ignored_players = tag_ids([
            &command.spectator_tag1,
            &command.spectator_tag2,
            &command.spectator_tag3,
            &command.spectator_tag4,
            &None,
            &None,
            &None,
            &None,
        ]);
        let manual_entries = tag_ids([
            &command.player_tag1,
            &command.player_tag2,
            &command.player_tag3,
            &command.player_tag4,
            &None,
            &None,
            &None,
            &None,
        ]);

        let response = divide_hub(
            bot,
            guild_id,
            user_id,
            command
                .game_mode
                .unwrap_or(GameModeOption::SacrificeTournament)
                .into(),
            sort_method,
            &manual_entries,
            &ignored_players,
        )
        .await?;
        Ok(Some(response))
    }
}

#[derive(CommandModel, CreateCommand, Debug)]
#[command(name = "backup-db", desc = "Backup DB to github")]
pub struct BackupDbCommand {}

impl BackupDbCommand {
    pub async fn handle(
        db: &Arc<RwLock<Db>>,
        github_config: &GithubConfig,
    ) -> Result<Option<InteractionResponseData>> {
        let json = db.read().await.to_json()?;
        upload_bytes_to_github(
            &json.into(),
            &github_config.owner,
            &github_config.repo,
            &github_config.path,
            &github_config.branch,
        )
        .await?;
        Ok(Some(text_response(format!(
            "https://raw.githubusercontent.com/{}/{}/{}/{}",
            github_config.owner, github_config.repo, github_config.branch, github_config.path
        ))))
    }
}

#[derive(CommandModel, CreateCommand, Debug)]
#[command(name = "restore-db-backup", desc = "Restore DB from github")]
pub struct RestoreDBBackupCommand {}

impl RestoreDBBackupCommand {
    pub async fn handle(
        db: &Arc<RwLock<Db>>,
        github_config: &GithubConfig,
    ) -> Result<Option<InteractionResponseData>> {
        let bytes = get_bytes_from_github(
            &github_config.owner,
            &github_config.repo,
            &github_config.path,
            &github_config.branch,
        )
        .await?;
        let restored = Db::from_json(&bytes)?;
        *db.write().await = restored;
        Ok(Some(text_response("DB restored from backup")))
    }
}
